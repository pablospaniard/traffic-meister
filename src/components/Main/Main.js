import React, { useContext } from 'react'
import { GridLoader } from 'react-spinners'
import isEmpty from 'lodash/isEmpty'
import styled from '@emotion/styled'

import { FlexContainer, FlexItem, Image } from '../UI'
import AppContext from '../../helpers/context'

const StyledLoader = styled.div`
  width: fit-content;
  margin: 20px auto;
`

const StyledFlexContainer = styled(FlexContainer)`
  border: 1px solid #ccc;
  width: 50%;
  margin: 5px auto;
  padding: 15px;
  border-radius: 4px;
  background-color: #607b7d;
  color: #e0e0e0;

  img {
    height: 100%;
    width: 200px;
    min-height: 50px;
  }

  @media (max-width: 1024px) {
    width: 100%;
  }
`

const StyledColorDiv = styled.div`
  display: inline-block;
  height: 20px;
  width: 20px;
  margin: 0 5px;
  background-color: ${({ color }) => color || 'inherit'};
  border: 1px solid #ccc;
`

const Main = () => {
  const { data, filteredData } = useContext(AppContext)

  return (
    <>
      {!isEmpty(data) ? (
        filteredData.map(item => (
          <StyledFlexContainer
            key={item.id}
            alignItems="center"
            data-testid="table"
          >
            <FlexItem flex="1">{item.type.toUpperCase()}</FlexItem>
            <FlexItem flex="1">{item.brand}</FlexItem>
            <FlexItem flex="1">
              {item.colors.map(color => (
                <StyledColorDiv key={color} color={color} />
              ))}
            </FlexItem>
            <Image src={item.img} alt={item.brand} />
          </StyledFlexContainer>
        ))
      ) : (
        <StyledLoader data-testid="loading">
          <GridLoader sizeUnit="px" size={10} color="#ccc" loading />
        </StyledLoader>
      )}
    </>
  )
}

export default Main
