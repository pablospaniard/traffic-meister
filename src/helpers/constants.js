export const TYPES = {
  CAR: 'car',
  TRAIN: 'train',
  PLANE: 'plane'
}

export const PROPS = {
  TYPES: 'type',
  BRANDS: 'brand',
  COLORS: 'color'
}
