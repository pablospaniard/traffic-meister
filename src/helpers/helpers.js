import isEmpty from 'lodash/isEmpty'

// eslint-disable-next-line import/prefer-default-export
export const filterData = (data, filters) => {
  const filteredData = isEmpty(filters) ? data : []
  const appliedTypes = !isEmpty(filters.types)
    ? filters.types.map(t => t.value)
    : []
  const appliedBrands = !isEmpty(filters.brands)
    ? filters.brands.map(t => t.value)
    : []
  const appliedColors = !isEmpty(filters.colors)
    ? filters.colors.map(t => t.value)
    : []
  data.forEach(item => {
    const values = Object.values(item)
    if (
      values.some(i => appliedTypes.includes(i) || appliedBrands.includes(i)) ||
      item.colors.some(c => appliedColors.includes(c))
    ) {
      filteredData.push(item)
    }
  })
  return filteredData
}
