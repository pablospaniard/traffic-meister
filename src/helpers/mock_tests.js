export const filters = {
  type: { label: 'car', value: 'Car' },
  brand: { label: 'yokohama', value: 'Yokohama' },
  color: { label: 'silver', value: 'silver' }
}
export const data = [
  {
    id: 1,
    type: 'car',
    brand: 'Bugatti Veyron',
    colors: ['red', 'black'],
    img:
      'https://upload.wikimedia.org/wikipedia/commons/thumb/c/c9/Bugatti_Veyron_16.4_%E2%80%93_Frontansicht_%281%29%2C_5._April_2012%2C_D%C3%BCsseldorf.jpg/520px-Bugatti_Veyron_16.4_%E2%80%93_Frontansicht_%281%29%2C_5._April_2012%2C_D%C3%BCsseldorf.jpg'
  },
  {
    id: 2,
    type: 'airplane',
    brand: 'Boeing 787 Dreamliner',
    colors: ['red', 'white', 'black', 'green'],
    img:
      'https://upload.wikimedia.org/wikipedia/commons/thumb/1/15/All_Nippon_Airways_Boeing_787-8_Dreamliner_JA801A_OKJ_in_flight.jpg/600px-All_Nippon_Airways_Boeing_787-8_Dreamliner_JA801A_OKJ_in_flight.jpg'
  },
  {
    id: 3,
    type: 'train',
    brand: 'USRA 0-6-6',
    colors: ['yellow', 'white', 'black'],
    img:
      'https://upload.wikimedia.org/wikipedia/en/thumb/a/a1/UP_4466_Neil916.JPG/600px-UP_4466_Neil916.JPG'
  }
]
