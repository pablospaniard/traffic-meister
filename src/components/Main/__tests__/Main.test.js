import React from 'react'
import { render } from '@testing-library/react'
import serializer from 'jest-emotion'

import Main from '../Main'
import AppContext from '../../../helpers/context'
import { data } from '../../../helpers/mock_tests'

expect.addSnapshotSerializer(serializer)

const tree = (
  <AppContext.Provider value={{ data, filteredData: data }}>
    <Main />
  </AppContext.Provider>
)

describe('Main component should works properly', () => {
  it('should match Snapshot', () => {
    const { container } = render(tree)
    expect(container.firstChild).toMatchSnapshot('Main_1')
  })

  it('should show Loader while data is empty', () => {
    const emptyTree = (
      <AppContext.Provider value={{ data: [], filteredData: data }}>
        <Main />
      </AppContext.Provider>
    )
    const { getByTestId } = render(emptyTree)
    getByTestId(/loading/i)
  })
})
