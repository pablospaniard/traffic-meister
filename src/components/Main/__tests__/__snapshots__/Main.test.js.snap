// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main component should works properly should match Snapshot: Main_1 1`] = `
.emotion-10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  border: 1px solid #ccc;
  width: 50%;
  margin: 5px auto;
  padding: 15px;
  border-radius: 4px;
  background-color: #607b7d;
  color: #e0e0e0;
}

.emotion-10 img {
  height: 100%;
  width: 200px;
  min-height: 50px;
}

@media (max-width:1024px) {
  .emotion-10 {
    width: 100%;
  }
}

.emotion-0 {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: block;
  margin: 0;
  width: 100%;
}

.emotion-4 {
  display: inline-block;
  height: 20px;
  width: 20px;
  margin: 0 5px;
  background-color: red;
  border: 1px solid #ccc;
}

.emotion-6 {
  display: inline-block;
  height: 20px;
  width: 20px;
  margin: 0 5px;
  background-color: black;
  border: 1px solid #ccc;
}

<div
  class="emotion-10 emotion-11"
  data-testid="table"
>
  <div
    class="emotion-0 emotion-1"
  >
    CAR
  </div>
  <div
    class="emotion-0 emotion-1"
  >
    Bugatti Veyron
  </div>
  <div
    class="emotion-0 emotion-1"
  >
    <div
      class="emotion-4 emotion-5"
      color="red"
    />
    <div
      class="emotion-6 emotion-5"
      color="black"
    />
  </div>
  <div
    class="imageloader imageloader-loading"
  >
    <div />
  </div>
</div>
`;
